core/src/main/java/io/questdb/griffin/engine/functions/str/SplitPartFunctionFactory.java

```diff
@@ -0,0 +1,210 @@
+        final Function strFunc = args.getQuick(0);
+        final Function delimiterFunc = args.getQuick(1);
+        final Function indexFunc = args.getQuick(2);
+        if (indexFunc.isConstant()) {
+            int index = indexFunc.getInt(null);
+            if (index == Numbers.INT_NaN) {
+                return StrConstant.NULL;
+            } else if (index == 0) {
+                return StrConstant.EMPTY;
+            } else {
+                return new SplitPartConstIndexFunction(strFunc, delimiterFunc, index);
+            }
+        }
+        return new SplitPartFunction(strFunc, delimiterFunc, indexFunc);
+    }
+
+    private static class SplitPartFunction extends AbstractSplitPartFunction implements TernaryFunction {
+        private final Function strFunc;
+        private final Function delimiterFunc;
+        private final Function indexFunc;
+
+        public SplitPartFunction(Function strFunc, Function delimiterFunc, Function indexFunc) {
+            super(strFunc, delimiterFunc);
+            this.strFunc = strFunc;
+            this.delimiterFunc = delimiterFunc;
+            this.indexFunc = indexFunc;
+        }
+
+        @Override
+        int getIndex(Record rec) {
+            return indexFunc.getInt(rec);
+        }
+
+        @Override
+        public Function getLeft() {
+            return strFunc;
+        }
+
+        @Override
+        public Function getCenter() {
+            return delimiterFunc;
+        }
+
+        @Override
+        public Function getRight() {
+            return indexFunc;
+        }
+    }
+
+    private static class SplitPartConstIndexFunction extends AbstractSplitPartFunction implements BinaryFunction {
+        private final Function strFunc;
+        private final Function delimiterFunc;
+        private final int index;
+
+        public SplitPartConstIndexFunction(Function strFunc, Function delimiterFunc, int index) {
+            super(strFunc, delimiterFunc);
+            this.strFunc = strFunc;
+            this.delimiterFunc = delimiterFunc;
+            this.index = index;
+        }
+
+        @Override
+        int getIndex(Record rec) {
+            return index;
+        }
+
+        @Override
+        public Function getLeft() {
+            return strFunc;
+        }
+
+        @Override
+        public Function getRight() {
+            return delimiterFunc;
+        }
+    }
+
+    private static abstract class AbstractSplitPartFunction extends StrFunction {
+        private final StringSink sink = new StringSink();
+        private final StringSink sinkB = new StringSink();
+
+        private final Function strFunc;
+        private final Function delimiterFunc;
+
+        public AbstractSplitPartFunction(Function strFunc, Function delimiterFunc) {
+            this.strFunc = strFunc;
+            this.delimiterFunc = delimiterFunc;
+        }
+
+        abstract int getIndex(Record rec);
```

Comment:
I see this pattern a lot in SQL function implementations:
```java
public class SplitPartFunctionFactory implements FunctionFactory {
    // ...

    @Override
    public Function newInstance(int position, ObjList<Function> args, IntList argPositions, CairoConfiguration configuration, SqlExecutionContext sqlExecutionContext) throws SqlException {
        final Function strFunc = args.getQuick(0);
        final Function delimiterFunc = args.getQuick(1);
        final Function indexFunc = args.getQuick(2);
        if (indexFunc.isConstant()) {
           int index = indexFunc.getInt(null);
            return new SplitPartConstIndexFunction(strFunc, delimiterFunc, index);
        }
        return new SplitPartFunction(strFunc, delimiterFunc, indexFunc);
    }

    private static class SplitPartFunction extends StrFunction implements TernaryFunction {
        private final StringSink sink = new StringSink();
        private final StringSink sinkB = new StringSink();

        private final Function strFunc;
        private final Function delimiterFunc;
        private final Function indexFunc;

        public SplitPartFunction(Function strFunc, Function delimiterFunc, Function indexFunc) {
            this.strFunc = strFunc;
            this.delimiterFunc = delimiterFunc;
            this.indexFunc = indexFunc;
        }

        @Override
        public CharSequence getStr(Record rec) {
            return getStr0(rec, sink);
        }

        @Override
        public CharSequence getStrB(Record rec) {
            return getStr0(rec, sinkB);
        }

        private StringSink getStr0(Record rec, StringSink sink) {
            var index = indexFunc.getInt(rec);
            // ...
        }

        // ...
    }

    private static class SplitPartConstIndexFunction extends StrFunction implements BinaryFunction {
        private final StringSink sink = new StringSink();
        private final StringSink sinkB = new StringSink();

        private final Function strFunc;
        private final Function delimiterFunc;
        private final int index;

        public SplitPartFunction(Function strFunc, Function delimiterFunc, int index) {
            this.strFunc = strFunc;
            this.delimiterFunc = delimiterFunc;
            this.index = index;
        }

        @Override
        public CharSequence getStr(Record rec) {
            return getStr0(rec, sink);
        }

        @Override
        public CharSequence getStrB(Record rec) {
            return getStr0(rec, sinkB);
        }

        private StringSink getStr0(Record rec, StringSink sink) {
            // use index
        }

        // ...
    }
}
```

This leads to a lot of code duplication between the default function implementation and the const function implementation, especially in the `str0(Record, StringSink)` functions.

To reduce code duplication, I created an abstract class (`AbstractSplitPartFunction`) with the shared `getStr(Record)` and `getStrB(Record)` methods, with an abstract method that returns the value (`getIndex(Record)`), i.e.
```java
public class SplitPartFunctionFactory implements FunctionFactory {
    // ...

    @Override
    public Function newInstance(int position, ObjList<Function> args, IntList argPositions, CairoConfiguration configuration, SqlExecutionContext sqlExecutionContext) throws SqlException {
        final Function strFunc = args.getQuick(0);
        final Function delimiterFunc = args.getQuick(1);
        final Function indexFunc = args.getQuick(2);
        if (indexFunc.isConstant()) {
           int index = indexFunc.getInt(null);
            return new SplitPartConstIndexFunction(strFunc, delimiterFunc, index);
        }
        return new SplitPartFunction(strFunc, delimiterFunc, indexFunc);
    }

    private static class SplitPartFunction extends AbstractSplitPartFunction implements TernaryFunction {
        private final StringSink sink = new StringSink();
        private final StringSink sinkB = new StringSink();
        private final Function strFunc;
        private final Function delimiterFunc;
        private final Function indexFunc;

        public SplitPartFunction(Function strFunc, Function delimiterFunc, Function indexFunc) {
            this.strFunc = strFunc;
            this.delimiterFunc = delimiterFunc;
            this.indexFunc = indexFunc;
        }

        @Override
        int getIndex(Record rec) {
            return indexFunc.getInt(rec);
        }

        // ...
    }

    private static class SplitPartConstIndexFunction extends AbstractSplitPartFunction implements BinaryFunction {
        private final Function strFunc;
        private final Function delimiterFunc;
        private final int index;

        public SplitPartFunction(Function strFunc, Function delimiterFunc, int index) {
            this.strFunc = strFunc;
            this.delimiterFunc = delimiterFunc;
            this.index = index;
        }

        @Override
        int getIndex(Record rec) {
            return index;
        }

        // ...
    }

    private static class AbstractSplitPartFunction extends StrFunction {
        private final StringSink sink = new StringSink();
        private final StringSink sinkB = new StringSink();

        private final Function strFunc;
        private final Function delimiterFunc;

        public AbstractSplitPartFunction(Function strFunc, Function delimiterFunc) {
            this.strFunc = strFunc;
            this.delimiterFunc = delimiterFunc;
        }

       abstract int getIndex(Record rec);

        @Override
        public CharSequence getStr(Record rec) {
            return getStr0(rec, sink);
        }

        @Override
        public CharSequence getStrB(Record rec) {
            return getStr0(rec, sinkB);
        }

        private StringSink getStr0(Record rec, StringSink sink) {
            var index = getIndex(rec);
            // ...
        }
    }
}
```